### ISSUE TITLE
Expressions with static type never 

### ISSUE THREAD

**Issue by hasithaa** [ballerina-platform/ballerina-spec#510]
Consider the following example.

```ballerina
function main() {
   fail();                  // 1.
   never value = fail();    // 2
   _ = fail();              // 3.
   () val = fail();         // 4.

   bar();                   // 5.
   _ = bar();               // 6.
}

function fail() returns never {
  error e = error("Argh...!!!");
  panic e;
}

function bar() { }
```

- Option 1 is allowed in the spec.
- Spec says, ` never is a subtype of nil`
- Technically, both 5 and 6 are the same.  
- `never` type can't have runtime representation, so case 2 is invalid IMO.

So the question is should we allow 3 and 4? 
If we allow 4, that means 2 also valid.

**Comment by jclark** [comment]
It wouldn't make sense to allow:

```
any x;
x = fail();
```

because the assignment to `x` would never happen, and so `x` would be uninitialized despite being assigned to.

I don't see 5 and 6 as the same: 6 is assigning to an anonymous variable whose value will never be used. Since we don't allow `x = fail()` for any x, I don't see why we should allow `_ = fail()`.

So I think think the rules should be:

- you can't have a variable of type `never`
- you can't have an intersection type equivalent to `never`
- you can't have an expression with static type `never`
- if you have a function with return type `never`, you can only call it in a call-stmt

**Comment by hasithaa** [comment]
I am +1 for the proposed design. If we are going with the above change, I think It should be a compile-time error if a function panics in all paths, and it doesn't have the never return type.

```ballerina
function fail() {      // This is an error
  panic error("Argh...!!!");
}

any x;
x = fail(); // Assignment never happens. 
```

**Comment by jclark** [comment]
That's a separable point, I think. It doesn't seem a good idea to me.

A function type is not merely describing the behaviour of a function body, but is also specifying a contract between caller and sender. When a function says "returns T", then contract is that if the function returns normally, the returned value belongs to T.

When I am writing a program, I often want to do:

```
public function something(int n) returns Something {
  panic error("Something not implemented yet!");
}
```

The function declaration is specifying a contract, and the implementation conforms to the contract.

We don't make this

```
public function foo(int n) returns int? {
  return n;
}
```

an compile error, even though foo can never return `()`. If that's not an error, then why should

```
public function foo(int n) returns () {
  panic error("");
}
```

be an error?

And what would you do about

```
public function foo(int n) returns () {
  _ = n/0;
  return ();
}
```

or

```
public function foo(int n) returns () {
   return foo();
}
```

**Comment by hasithaa** [comment]
Yes, I agree with your point. Sometimes we need to keep the same function type to make structural equivalency to work.

**Comment by dulajdilshan** [comment]
Just for the clarification.

If we are not assigning a value for x. Is there any point in having even an optional field with 'never' type ?

```
type Person record {
    never x?;
}
```

**Comment by jclark** [comment]
@dulajdilshan Yes, because it allows us to describe a record type that does _not_ have a specific field. This means that `{x: 1, ...p}` where `p` has type `Person` will not have any duplicate fields.

**Comment by pubudu91** [comment]
A couple of questions regarding `never`. 
- Should call expressions of type `never` be allowed in start-action? 
```ballerina
future<never> f = start foo();
```
If not, should the declaration `future<never> f;` be disallowed?
- Similar question regarding streams. Does `stream<never> st;` make sense/be allowed?

**Comment by jclark** [comment]
It's certainly useful to be able to apply `start` to functions that return `never`, so the type should be `future<never>`. You need to be careful about a wait-action applied to such a future.

`stream<never>` doesn't make any sense, but `stream<T,never>` does: it's an unbounded stream.

**Comment by jclark** [comment]
Object type with parameter of type `never` makes sense (see #615).

**Comment by hasithaa** [comment]
Another Interesting use case with never (but not useful IMO)
```ballerina
type X record {| |};
type Y record {|  never bar; |};
```
In both Xand Y have the same value space, which is `{}`, hence, both X and Y considered structurally equivalent

**Comment by jclark** [comment]
We ought to be able to say:

```
function bye() returns never {
   panic error("bye, bye");
}

public function main() returns error? {
   return trap bye();
}
```

@hasithaa What do you think?

**Comment by jclark** [comment]
Another interesting case is here: https://github.com/ballerina-platform/ballerina-spec/issues/774#issuecomment-793214385

**Comment by jclark** [comment]
What I've done so far is as follows.

For expressions, never is allowed only when both kind of expression and context of expression permit never. Kinds of expression permitted to be never are:

- function call
- method call (but not client remote method call)

Context where expression of type never is permitted:

- call-stmt
- start-action
- trap-expr

For variables, `never` disallowed as type-descriptor as follows:

- parameters: required-param and defaultable-param
- local variable declaration without initializer
- module variable declaration without initializer
- module const declaration
- typed-binding-pattern (includes variables with initializers)

I think we should also disallow never as a *required* member of structure types e.g. `record { never x; }`
or `[never]`. This doesn't disallow `never` for optional or `...` member types.

@hasithaa, @MaryamZi Please review

**Comment by MaryamZi** [comment]
If we are to disallow `never` as the type of a required member of a structure, I guess this restriction would apply to the type descriptor? Which would make the following example (based on https://github.com/ballerina-platform/ballerina-spec/issues/774) result in an error even though `x.iterator().next()` will return `()`?

```ballerina
public function main() {
    map<never> x = {};
    record {| never value; |}? y = x.iterator().next(); // Error for `record {| never value; |}`
}
```

If the above will result in an error, what about when such a type is inferred? 

```ballerina
public function main() {
    map<never> x = {};
    var y = x.iterator().next();
}
```

**Comment by jclark** [comment]
@MaryamZi That's a good point about the result of an iterator `next` method.

Given that, I think it's better not to disallow `never` as a required member. That means that if you have a structure type with a required `never` member, that structure type will be equivalent to never.

We can then combine this with the distinction I made in the fix to #737 between two kinds of use of typed-binding-pattern:

- single use (followed by `=` plus `on fail`)
- iterative use (followed by `in`)

and refine the prohibition on variables having type `never` to apply only in the single use case.

Then we can make #774 work for `map<never>`  (or `xml<never>` or `[]`) smoothly with no special casing needed

- the return type of the iterator is `record {| never value; |}?` which is equivalent to `never?`, which is `never|()`, which is `()`
- this means the iterator is known at compile time to iterate zero times
- the inferred type of the variable is `never`, but the no value is ever bound to the typed-binding-pattern
